File: 2024/day20.py
import os
import sys
import math
import numpy as np
from collections import Counter
from itertools import permutations

test = """\
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############"""

day = os.path.splitext(os.path.basename(__file__))[0]

TEST = 'test' in sys.argv
DEBUG = 'debug' in sys.argv

if TEST:
    data = test
else:
    data = open(day+'.txt').read()

data = data.splitlines()

WIDTH = len(data[0])
HEIGHT = len(data)

walls = set()
start = None
end = None
for y,row in enumerate(data):
    for x,c in enumerate(row):
        if c=='#':
            walls.add( (x,y))
        elif c=='S':
            start = (x,y)
        elif c=='E':
            end = (x,y)

dirs = [(-1,0),(0,-1),(1,0),(0,1)]

# There is only one path through.  Compute the distance from start for each step.

def makemap(walls):
    mapx = {}
    point = start
    while point != end:
        mapx[point] = len(mapx)
        for dx,dy in dirs:
            x0=point[0]+dx
            y0=point[1]+dy
            if (x0,y0) not in walls and (x0,y0) not in mapx:
                point = (x0,y0)
                break
    mapx[end] = len(mapx)
    return mapx

path = makemap(walls)

def part2(path, cheat, crit):
    sumx = 0
    counter = Counter()
    apath = np.array(list(path.keys()))
    # For each pair of points, how much would be gained by shortcutting them?
    for i,a in enumerate(path.keys()):
        mandist = np.abs(apath[i+1:,0]-a[0]) + np.abs(apath[i+1:,1]-a[1])
        for n in np.argwhere(mandist <= cheat):
            # I don't understand why I need the [0] here.  apath[i+1+n] should return
            # [11,22] but it returns [[11,22]].
            op = tuple(apath[i+1+n][0])
            ad = path[op] - path[a]
            gain = ad-mandist[n][0]
            if gain >= crit:
                counter[gain] += 1
                sumx += 1
    if DEBUG:
        print(counter)
    return sumx

print("Part 1:", part2(path,  2, 20 if TEST else 100))
print("Part 2:", part2(path, 20, 50 if TEST else 100))

File: 2024/day20.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false

var test string = `
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############`[1:]

//go:embed day20.txt
var live string

type Point = tools.Point

type PointSet = map[Point]bool

var WIDTH int = -1
var HEIGHT int = -1
var start Point
var finish Point
var directions []Point = []Point{{-1, 0}, {0, -1}, {0, 1}, {1, 0}}

// There is only one path through.  Compute the distance from start for each step.

func makemap(grid []string) map[Point]int {
	mapx := make(map[Point]int)
	point := start
	for point != finish {
		mapx[point] = len(mapx) + 1
		for _, dir := range directions {
			npt := point.Add(dir) 
			if grid[npt.Y][npt.X] != '#' && mapx[npt] == 0 {
				point = npt
				break
			}
		}
	}

	mapx[finish] = len(mapx) + 1
	return mapx
}

func part2(path map[Point]int, cheat int, crit int) int {
	sum := 0

	// For each pair of points, how much would be gained by shortcutting them?

	for k1, v1 := range path {
		for k2, v2 := range path {
			if v1 >= v2 {
				continue
			}
			// Compute Manhattan distance.

			manhattan := tools.Mandist(k1, k2)
			if manhattan <= cheat {
				actual := v2 - v1
				gain := actual - manhattan
				if gain >= crit {
					sum++
				}
			}
		}
	}
	return sum
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	data := strings.Split(input, "\n")

	WIDTH = len(data[0])
	HEIGHT = len(data)

	for y, row := range data {
		for x, c := range row {
			if c == 'S' {
				start = Point{x, y}
			} else if c == 'E' {
				finish = Point{x, y}
			}
		}
	}

	path := makemap(data)

	c1, c2 := 100, 100
	if TEST {
		c1, c2 = 20, 50
	}

	fmt.Println("Part 1:", part2(path, 2, c1))
	fmt.Println("Part 2:", part2(path, 20, c2))
}
